aten/src/ATen/native/sparse/SparseTensor.cpp

```diff
@@ -156,31 +179,39 @@ SparseTensor new_with_tensor_and_size_sparse(const LongTensor& indices, const Te
 
   int64_t sparseDims = indices.size(0);
   int64_t denseDims = values.dim() - 1;
-  AT_CHECK(sizes.size() == sparseDims + denseDims, "number of dimensions must be sparseDims (", sparseDims, ") + denseDims (", denseDims, "), but got ", sizes);
+  AT_CHECK(sizes.size() == sparseDims + denseDims, "number of dimensions must be sparseDims (", sparseDims, ") + denseDims (", denseDims, "), but got ", sizes.size());
+
+  // Check to make sure all indices are within the boundaries of `sizes`
+  // NOTE: We don't need to check this for scalar sparse tensors, which
+  // has sizes.size() == sparseDims == denseDims == 0
+  if (sizes.size() > 0) {
+    LongTensor max_indices = std::get</* values */ 0>(indices.max(/* dim */ 1, /* keepdim */ false));
+    LongTensor cpu_max_indices;
+    if (max_indices.is_cuda()) {
+      cpu_max_indices = at::CPU(kLong).copy(max_indices);
+    } else {
+      cpu_max_indices = max_indices;
+    }
+    auto cpu_max_indices_accessor = cpu_max_indices.accessor<int64_t, 1>();
+    for (int64_t d = 0; d < sparseDims; d++) {
+      // NB: This used to sync ndim times to access each entry; now we copy
+      // everything to CPU first and then access it.
+      int64_t max_index_in_dim = cpu_max_indices_accessor[d];
+      int64_t dim_size = sizes[static_cast<size_t>(d)];
+      AT_CHECK(max_index_in_dim < dim_size,
+               "sizes is inconsistent with indices: for dim ", d, ", size is ", dim_size, " but found index ", max_index_in_dim);
+    }
 
-  LongTensor max_indices = std::get</* values */ 0>(indices.max(/* dim */ 1, /* keepdim */ false));
-  LongTensor cpu_max_indices;
-  if (max_indices.is_cuda()) {
-    cpu_max_indices = at::CPU(kLong).copy(max_indices);
-  } else {
-    cpu_max_indices = max_indices;
-  }
-  auto cpu_max_indices_accessor = cpu_max_indices.accessor<int64_t, 1>();
-  for (int64_t d = 0; d < sparseDims; d++) {
-    // NB: This used to sync ndim times to access each entry; now we copy
-    // everything to CPU first and then access it.
-    int64_t max_index_in_dim = cpu_max_indices_accessor[d];
-    int64_t dim_size = sizes[static_cast<size_t>(d)];
-    AT_CHECK(max_index_in_dim < dim_size,
-             "sizes is inconsistent with indices: for dim ", d, ", size is ", dim_size, " but found index ", max_index_in_dim);
-  }
-  for (int64_t d = 0; d < denseDims; d++) {
-    int64_t values_size = values.size(d+1);
-    int64_t specified_size = sizes[static_cast<size_t>(sparseDims + d)];
-    AT_CHECK(values_size <= specified_size,
-             "values and sizes are inconsistent: sizes[", d + sparseDims, "] is ", specified_size,
-             " but values.size(", d + 1, ") is ", values_size);
+    // Check to make sure the dense dimensions match between `values` and `sizes`
+    for (int64_t d = 0; d < denseDims; d++) {
```

Comment:
does it make sense to unify all the checking with where you do it above (in SparseTensorImpl.cpp?)